### ISSUE TITLE
master: MPI_Irecv seems not to be thread safe?

### ISSUE THREAD

**Issue by thananon** [open-mpi/ompi#2921]
This is the message I get from multithreaded injection rate benchmark. The benchmark post multiple Irecv on the receiver side (multiple threads at a time). 

This happens on both [Artem's benchmark](https://github.com/artpol84/poc) and my own benchmark and only happens on Open MPI and has been happening for a while. (I tested with Intel MPI and everything is ok)

```
[mpi002:27049] *** An error occurred in MPI_Irecv
[mpi002:27049] *** reported by process [56688641,1]
[mpi002:27049] *** on communicator MPI COMMUNICATOR 6 DUP FROM 0
[mpi002:27049] *** MPI_ERR_INTERN: internal error
[mpi002:27049] *** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
[mpi002:27049] ***    and potentially your MPI job)
```

It happens with different BTLs and I think @jsquyres can confirm this.

It's not happening all the time. This will happens more often if multiple threads are bounded to the same core. This suggests something happens between context switching?

**Comment by jsquyres** [comment]
Just to be clear: this is on master?  v2.0.x?  v2.x?

**Comment by thananon** [comment]
@jsquyres I have been seeing this before we even released 2.0.0. I thought it was something funny with the benchmark until I tried with different MPI and different benchmark.

**Comment by hjelmn** [comment]
@thananon Arm, are you going to look at this or would you like me to take a look?

**Comment by thananon** [comment]
@hjelmn If you have cycle for this, can you please check it out? I am in the middle of other task but I will come back and take a look at this when I have time.

**Comment by hjelmn** [comment]
Happens because we somehow run out of fragments. Looking into it.

**Comment by hjelmn** [comment]
Keeping this open until merged into v3.0.x, v3.1.x, and v4.0.x.
